### PR TITLE
asserts,seed: split handling of essential/not essential model snaps

### PR DESCRIPTION
**Pull request by pedronis**
this splits the handling of essential and not essential model snaps
by having model.EssentialSnaps and model.SnapsWithoutEssential
instead of AllSnaps

notice that the order of snaps is still relevant in seedwriter and
seed, but is now managed more explicitly:

 * system snap first (if implicit)
 * essential snaps (in the EssentialSnaps order)
 * not essential snaps (in SnapsWithoutEssential order)

### CODE REVIEW

File: asserts/model.go [hunk 1]
```diff
@@ -486,19 +486,29 @@ func (mod *Model) Store() string {
...
+// They are always returned according to this order with some skipped
+// if not mentioned: snapd, kernel, boot base, gadget.
+func (mod *Model) EssentialSnaps() []*ModelSnap {
+	return mod.allSnaps[:mod.numEssentialSnaps]
+}
+
+// SnapsWithoutEssential returns all the snaps listed by the model
+// without any of the essential snaps (as returned by EssentialSnaps).
+// They are returned in the order of mention by the model.
+func (mod *Model) SnapsWithoutEssential() []*ModelSnap {
```
Comment (stolowski) [inline asserts/model.go:510]: Should this be named NonEssentialSnaps?

Comment (pedronis) [reply asserts/model.go:510]: maybe, I don't like the connotation of that, it really sounds like they can be omitted

File: asserts/model_test.go [hunk 2]
```diff
@@ -195,14 +198,22 @@ func (mods *modelSuite) TestDecodeOK(c *C) {
...
 	// essential snaps excluded
-	c.Check(model.RequiredNoEssentialSnaps(), DeepEquals, reqSnaps[3:])
+	noEssential := naming.NewSnapSet(model.RequiredNoEssentialSnaps())
+	for _, e := range essentialSnaps {
+		c.Check(noEssential.Contains(e), Equals, false)
+	}
+	for _, s := range snaps {
+		c.Check(noEssential.Contains(s), Equals, true)
+	}
+
```
Comment (stolowski) [inline asserts/model_test.go:216]: Could we also check that reqSnaps and noEssential were exhausted by above loops? e.g. by simple length checks after the loops (len(reqSnaps) == len(essential)+len(snaps) etc.)?

File: asserts/model_test.go [hunk 3]
```diff
@@ -551,14 +565,21 @@ func (mods *modelSuite) TestClassicDecodeOK(c *C) {
...
 	}
 	// gadget excluded
-	c.Check(model.RequiredNoEssentialSnaps(), DeepEquals, reqSnaps[1:])
+	noEssential := naming.NewSnapSet(model.RequiredNoEssentialSnaps())
+	for _, e := range essentialSnaps {
+		c.Check(noEssential.Contains(e), Equals, false)
+	}
+	for _, s := range snaps {
+		c.Check(noEssential.Contains(s), Equals, true)
+	}
```
Comment (stolowski) [inline asserts/model_test.go:582]: Same about also checking that all reqSnaps / noEssential were actually verified.

**Pull request merged**